sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,277 @@
+ *
+ * }</pre>
+ */
+public class CassandraIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CassandraIO.class);
+
+  private CassandraIO() {}
+
+  public static <T> Read<T> read() {
+    return new AutoValue_CassandraIO_Read.Builder<T>().build();
+  }
+
+  /**
+   * A {@link PTransform} to read data from Apache Cassandra. See {@link CassandraIO} for more
+   * information on usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+
+    @Nullable abstract List<String> hosts();
+    @Nullable abstract Integer port();
+    @Nullable abstract String keyspace();
+    @Nullable abstract String table();
+    @Nullable abstract Class<T> entity();
+    @Nullable abstract Coder<T> coder();
+    @Nullable abstract CassandraService<T> cassandraService();
+    abstract Builder<T> builder();
+
+    public Read<T> withHosts(List<String> hosts) {
+      checkArgument(hosts != null, "CassandraIO.read().withHosts(hosts) called with null hosts");
+      checkArgument(hosts.isEmpty(), "CassandraIO.read().withHosts(hosts) called with empty "
+          + "hosts list");
+      return builder().setHosts(hosts).build();
+    }
+
+    public Read<T> withPort(int port) {
+      checkArgument(port > 0, "CassandraIO.read().withPort(port) called with invalid port "
+          + "number (" + port + ")");
+      return builder().setPort(port).build();
+    }
+
+    public Read<T> withKeyspace(String keyspace) {
+      checkArgument(keyspace != null, "CassandraIO.read().withKeyspace(keyspace) called with "
+          + "null keyspace");
+      return builder().setKeyspace(keyspace).build();
+    }
+
+    public Read<T> withTable(String table) {
+      checkArgument(table != null, "CassandraIO.read().withTable(table) called with null table");
+      return builder().setTable(table).build();
+    }
+
+    public Read<T> withEntity(Class<T> entity) {
+      checkArgument(entity != null, "CassandraIO.read().withEntity(entity) called with null "
+          + "entity");
+      return builder().setEntity(entity).build();
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
+      checkArgument(coder != null, "CassandraIO.read().withCoder(coder) called with null coder");
+      return builder().setCoder(coder).build();
+    }
+
+    public Read<T> withCassandraService(CassandraService<T> cassandraService) {
+      checkArgument(cassandraService != null, "CassandraIO.read().withCassandraService(service)"
+          + " called with null service");
+      return builder().setCassandraService(cassandraService).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new CassandraSource<T>(this, null)));
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(hosts() != null || cassandraService() != null,
```

Comment:
Would be more clear to say: if (cassandraService() == null) { checkBlah, checkBlah }